src/classes/trackers.js

```diff
@@ -1,290 +1,498 @@
+        }
+        const specialProtocols = ['chrome-extension:', 'chrome-devtools:', 'chrome-search:', 'about:', 'file:', 'javascript:', 'data:']
+        if (urlObj) {
+            if (specialProtocols.includes(urlObj.protocol)) {
+                return true
             }
         }
+        return false
+    }
 
-        /*
-         * Pull subdomains off of the reqeust rule and look for a matching tracker object in our data
-         */
-        findTracker (requestData) {
-            const urlList = Array.from(requestData.urlToCheckSplit)
+    /**
+     * @param {string} urlToCheck
+     * @param {string} siteUrl
+     * @param {RequestExpression} request
+     * @returns {TrackerData | null}
+     */
+    getTrackerData (urlToCheck, siteUrl, request) {
+        if (!this.entityList || !this.trackerList) {
+            throw new Error('tried to detect trackers before rules were loaded')
+        }
 
-            while (urlList.length > 1) {
-                const trackerDomain = urlList.join('.')
-                urlList.shift()
+        if (this.isSpecialURL(urlToCheck)) {
+            return null
+        }
 
-                const matchedTracker = this.trackerList[trackerDomain]
-                if (matchedTracker) {
-                    return matchedTracker
+        let fromCname
+        let requestData = this.getRequestData(urlToCheck, siteUrl, request)
+        if (!requestData) {
+            return null
+        }
+        const requestOwner = this.findTrackerOwner(requestData.urlToCheckDomain)
+        const websiteOwner = this.findWebsiteOwner(requestData)
+
+        const firstParty = (requestOwner && websiteOwner) ? requestOwner === websiteOwner : requestData.siteDomain === requestData.urlToCheckDomain
+        const fullTrackerDomain = requestData.urlToCheckSplit.join('.')
+
+        // finds a tracker definition by iterating over the whole trackerList and finding the matching tracker.
+        let tracker = this.findTracker(requestData)
+
+        if (!tracker) {
+            // if request doesn't have any rules associated with it, we should check if it's a CNAMEed tracker
+            const cnameResolution = this.resolveCname(urlToCheck)
+            fromCname = cnameResolution.fromCname
+            const cnameRequestData = this.getRequestData(cnameResolution.finalURL, siteUrl, request)
+            if (!cnameRequestData) {
+                return null
+            }
+            tracker = this.findTracker(cnameRequestData)
+            if (tracker) {
+                requestData = cnameRequestData
+            } else {
+                if (firstParty) {
+                    return null
+                }
+                const owner = {
+                    name: requestOwner || requestData.urlToCheckDomain || 'unknown',
+                    displayName: requestOwner || requestData.urlToCheckDomain || 'Unknown'
+                }
+                /** @type {TrackerObj} */
+                const tracker = {
```

Comment:
Nit: This is explicitly not a tracker. Should we call it something else? Maybe `const result = {`. We may also want to rename the class to `ResultObj`.

Reply:
Ignoring as this would change the output property which would require extension changes.